cms/djangoapps/contentstore/management/commands/backfill_orgs_and_org_courses.py

```diff
@@ -0,0 +1,168 @@
+
+
+class Command(BaseCommand):
+    """
+    Back-populate edx-organizations models from existing courses & content libraries.
+
+    Before the Lilac open release, Open edX instances by default did not make
+    use of the models in edx-organizations.
+    In Lilac and beyond, the edx-organizations models are enabled globally.
+
+    This command exists to migrate pre-Lilac instances that did not enable
+    `FEATURES['ORGANIZATIONS_APP']`.
+    It automatically creates all missing Organization and OrganizationCourse
+    instances based on the course runs in the system (loaded from CourseOverview)
+    and the V1 content libraries in the system (loaded from the Modulestore).
+
+    Organizations created by this command will have their `short_name` and
+    `name` equal to the `org` part of the library/course key that triggered
+    their creation. For example, given an Open edX instance with the course run
+    `course-v1:myOrg+myCourse+myRun` but no such Organization with the short name
+    "myOrg" (case-insensitive), this command will create the following
+    organization:
+        > Organization(
+        >     short_name='myOrg',
+        >     name='myOrg',
+        >     description=None,
+        >     logo=None,
+        >     active=True,
+        > )
+    """
+    help = dedent(__doc__).strip()
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--apply',
+            action='store_true',
+            help="Apply backfill to database (instead of just showing the changes)."
+        )
+        parser.add_argument(
+            '--dry',
+            action='store_true',
+            help="Apply backfill to database (instead of just showing the changes)."
+        )
+
+    def handle(self, *args, **options):
+
+        # Find `orgs` and `org_coursekey_pairs` to be bulk-added.
+        # The `sorted` calls aren't strictly necessary, but they'll help make this
+        # function more deterministic in case something goes wrong.
+        orgslug_coursekey_pairs = find_orgslug_coursekey_pairs()
+        orgslug_library_pairs = find_orgslug_library_pairs()
+        orgslugs = (
+            {orgslug for orgslug, _ in orgslug_coursekey_pairs} |
+            {orgslug for orgslug, _ in orgslug_library_pairs}
+        )
+        orgs = [
+            {"short_name": orgslug, "name": orgslug}
+            for orgslug in sorted(orgslugs)
+        ]
+        org_coursekey_pairs = [
+            ({"short_name": orgslug}, coursekey)
+            for orgslug, coursekey in sorted(orgslug_coursekey_pairs)
+        ]
+
+        # Note that edx-organizations code will handle:
+        # * Not overwriting existing organizations.
+        # * Skipping duplicates, based on the short name (case-insensiive).
+
+        # Start with a dry run.
+        # This will log to the user which orgs/org-courses will be created/reactivated.
+        organizations_api.bulk_add_organizations(
```

Comment:
Is there a reason to do the dry run even if we're about to do the real run?  I feel like what I normally see is that you can use `--dry` to test what will happen, but then after you're satisfied with it, you just run the real thing and it doesn't perform _another_ dry run before that. 

Seems like this whole latter half of the `handle` function could be simplified:

1. Use `--dry` option to decide if `dry_run` should be True or False for `bulk_add_organizations` and `bulk_add_organization_courses`
2. Call 'em both with the value.

Then the user's responsible for choosing whether they want to use `--dry` or not, and "apply" is just assumed if `--dry` isn't present.

I think the above is moot if `should_apply_changes` is a pattern we've used elsewhere - the consistency is probably more important, then.  And I admit it feels like a nice DX to ask them to confirm, but it is a bit different/redundantly safe than I would expect.

Reply:
@davidjoy The "prompt if not specified by arguments" pattern here seemed in line with command-line administration tools I use day-to-day. For example, Debian's package manager `apt` will prompt for all major changes, unless you specify `--yes` (assume "yes" as input) or `--dry-run` (assume "no" as input). I see other commands in edx-platform that have yes/no prompts.

The database effect of this command is very large -- it potentially writes thousands of rows into production tables. That seemed a bit scary to do unprompted with just this command:
```
./manage.py cms backfill_orgs_and_org_courses
```
If an operator is specifying this, though:
```
./manage.py cms backfill_orgs_and_org_courses --apply
```
then it feels more likely that they know what they're getting themself into.

I'll admit that the control flow is a bit wonky in the `./manage.py cms backfill_orgs_and_org_courses --apply` case; it does an unnecessary dry run before doing the real run. I can try to straighten that out a bit.

Reply:
@davidjoy Pushed a new commit that simplifies the control flow a bit, but keeps the functionality the same.